docs/2024/puzzles/day18.md

```diff
@@ -1,11 +1,181 @@
+
+```scala
+extension (walls: Set[Vec2i])
+  def search(gridSize: Int): Option[List[Vec2i]] =
+    def reconstructPath(cameFrom: Map[Vec2i, Vec2i], p: Vec2i): List[Vec2i] = {
+      val totalPath = mut.ListBuffer[Vec2i](p)
+      var current = p
+      while (cameFrom.contains(current)) {
+        current = cameFrom(current)
+        totalPath.prepend(current)
+      }
+      totalPath.toList
+    }
+
+    val start = Vec2i(0, 0)
+    val goal = Vec2i(gridSize - 1, gridSize - 1)
+    val cameFrom = mut.HashMap[Vec2i, Vec2i]()
+
+    val dist = mut.HashMap(start -> 0d)
+
+    val q = mut.PriorityQueue(start -> 0d)(using Ordering.by[(Vec2i, Double), Double](_._2).reverse)
+
+    while q.nonEmpty do
+      val (current, score) = q.dequeue()
+
+      if current == goal then
+        return Some(reconstructPath(cameFrom.toMap, current))
+
+      for neighbor <- current.cardinalNeighbors.filter(it => it.isContainedIn(gridSize, gridSize) && !walls.contains(it)) do
+        val alt = score + 1d
+        if alt < dist.getOrElse(neighbor, Double.PositiveInfinity) then
+          cameFrom(neighbor) = current
+          dist(neighbor) = alt
+          q.addOne(neighbor -> alt)
+
+    None
+```
+
+Now part 1 is just running this with the correct inputs:
+```scala
+def part1(str: String): Int =
+  val (size, take, walls) = parse(str)
+  walls.take(take).toSet.search(size).get.size - 1
+```
+
+(We subtract 1 here to get the number of moves made, instead of the number of tiles.)
+
+## Part 2
+
+Part 2 is finding the point where the path is impossible to reach. A naive approach would be a linear search, and with the input size it is perfectly acceptable
+with only taking 2 seconds, but a better approach is a binary search.
+
+All of part 2's code is entirely within it's code segment:
+
+```scala
+def part2(str: String): Vec2i =
+  val (size, take, walls) = parse(str)
+  Iterator.iterate(take -> walls.size): (i0, i1) =>
+    if walls.take((i0 + i1) / 2).toSet.search(size).isEmpty then i0 -> (i0 + i1) / 2 else (i0 + i1) / 2 + 1 -> i1
+  .flatMap: (i0, i1) =>
+    Option.when(i0 == i1)(walls(i0 - 1))
+  .next()
+```
+
+This code will need some explaining. The iterate function is iterating through a binary search: it picks the midpoint of its two endpoints
+and searches the path with that amount of bytes fallen. If the result is None, meaning there is no valid path, then it pins its higher end point
+to that midpoint and repeats. If it's Some, meaning there is a valid path, then it pins the lower endpoint to the midpoint + 1. Doing this means you can
+invalidate large swathes of combinations from having to be tested. Before rewriting my code to use binary search, it took around 2 seconds. This code takes
+around 50 milliseconds, so that's a 40x speedup just from using binary search.
+
+The flatMap(...).next() just finds the first point where i0 == i1, or where the binary search has completed. Because we've been taking, and not indexing, we have
```

Comment:
```suggestion
The `flatMap(...).next()` just finds the first point where `i0 == i1`, or where the binary search has completed. Because we've been taking, and not indexing, we have
```